cpp/src/neighbors/nn_descent_c.cpp

```diff
@@ -0,0 +1,163 @@
+    return new cuvs::neighbors::nn_descent::index<IdxT>(std::move(index));
+  } else if (cuvs::core::is_dlpack_host_compatible(dataset)) {
+    using dataset_type = raft::host_matrix_view<T const, int64_t, raft::row_major>;
+    auto dataset       = cuvs::core::from_dlpack<dataset_type>(dataset_tensor);
+    auto index         = cuvs::neighbors::nn_descent::build(*res_ptr, build_params, dataset, graph);
+    return new cuvs::neighbors::nn_descent::index<IdxT>(std::move(index));
+  } else {
+    RAFT_FAIL("dataset must be accessible on host or device memory");
+  }
+}
+}  // namespace
+
+extern "C" cuvsError_t cuvsNNDescentIndexCreate(cuvsNNDescentIndex_t* index)
+{
+  return cuvs::core::translate_exceptions([=] { *index = new cuvsNNDescentIndex{}; });
+}
+
+extern "C" cuvsError_t cuvsNNDescentIndexDestroy(cuvsNNDescentIndex_t index_c_ptr)
+{
+  return cuvs::core::translate_exceptions([=] {
+    auto index = *index_c_ptr;
+    if ((index.dtype.code == kDLUInt) && (index.dtype.bits == 32)) {
+      auto index_ptr = reinterpret_cast<cuvs::neighbors::nn_descent::index<uint32_t>*>(index.addr);
+      delete index_ptr;
+    } else {
+      RAFT_FAIL(
+        "Unsupported nn-descent index dtype: %d and bits: %d", index.dtype.code, index.dtype.bits);
+    }
+    delete index_c_ptr;
+  });
+}
+
+extern "C" cuvsError_t cuvsNNDescentBuild(cuvsResources_t res,
+                                          cuvsNNDescentIndexParams_t params,
+                                          DLManagedTensor* dataset_tensor,
+                                          DLManagedTensor* graph_tensor,
+                                          cuvsNNDescentIndex_t index)
+{
+  return cuvs::core::translate_exceptions([=] {
+    index->dtype.code = kDLUInt;
+    index->dtype.bits = 32;
+
+    auto dtype = dataset_tensor->dl_tensor.dtype;
+
+    if ((dtype.code == kDLFloat) && (dtype.bits == 32)) {
+      index->addr = reinterpret_cast<uintptr_t>(
+        _build<float, uint32_t>(res, *params, dataset_tensor, graph_tensor));
+    } else if ((dtype.code == kDLFloat) && (dtype.bits == 16)) {
+      index->addr = reinterpret_cast<uintptr_t>(
+        _build<half, uint32_t>(res, *params, dataset_tensor, graph_tensor));
+    } else if ((dtype.code == kDLInt) && (dtype.bits == 8)) {
+      index->addr = reinterpret_cast<uintptr_t>(
+        _build<int8_t, uint32_t>(res, *params, dataset_tensor, graph_tensor));
+    } else if ((dtype.code == kDLUInt) && (dtype.bits == 8)) {
+      index->addr = reinterpret_cast<uintptr_t>(
+        _build<uint8_t, uint32_t>(res, *params, dataset_tensor, graph_tensor));
+    } else {
+      RAFT_FAIL("Unsupported nn-descent dataset dtype: %d and bits: %d", dtype.code, dtype.bits);
+    }
+  });
+}
+
+extern "C" cuvsError_t cuvsNNDescentIndexParamsCreate(cuvsNNDescentIndexParams_t* params)
+{
+  return cuvs::core::translate_exceptions([=] {
+    *params = new cuvsNNDescentIndexParams{.metric                    = L2Expanded,
```

Comment:
If we're going to duplicate the params structs for now, can we at least use constants to specify the defaults? I can see a lot of confusion and long debugging sessions happening as a result of these not getting synchronized between the C++ and C APIs.

Reply:
how about we have the canonical defaults in the c++ struct - and then pull out of there ?  Something like : https://github.com/rapidsai/cuvs/pull/635/commits/19deed988e2c2b8474d076f4c867ed5210f7a0e7